moo.js

```diff
@@ -335,40 +335,27 @@
   }
 
   function keywordTransform(map) {
-    var reverseMap = Object.create(null)
-    var byLength = Object.create(null)
+
+    // Use a JavaScript Map to map keywords to their corresponding token type
+    // unless Map is unsupported, then fall back to using an Object:
+    var isMap = typeof Map !== 'undefined'
+    var reverseMap = isMap ? new Map : Object.create(null)
+
     var types = Object.getOwnPropertyNames(map)
     for (var i = 0; i < types.length; i++) {
       var tokenType = types[i]
       var item = map[tokenType]
       var keywordList = Array.isArray(item) ? item : [item]
       keywordList.forEach(function(keyword) {
-        (byLength[keyword.length] = byLength[keyword.length] || []).push(keyword)
         if (typeof keyword !== 'string') {
           throw new Error("keyword must be string (in keyword '" + tokenType + "')")
         }
-        reverseMap[keyword] = tokenType
+        isMap ? reverseMap.set(keyword, tokenType) : reverseMap[keyword] = tokenType
       })
     }
-
-    // fast string lookup
-    // https://jsperf.com/string-lookups
-    function str(x) { return JSON.stringify(x) }
-    var source = ''
-    source += 'switch (value.length) {\n'
-    for (var length in byLength) {
-      var keywords = byLength[length]
-      source += 'case ' + length + ':\n'
-      source += 'switch (value) {\n'
-      keywords.forEach(function(keyword) {
-        var tokenType = reverseMap[keyword]
-        source += 'case ' + str(keyword) + ': return ' + str(tokenType) + '\n'
-      })
-      source += 'default: return\n'
-      source += '}\n'
+    return function(k) {
+      return isMap ? reverseMap.get(k) : reverseMap[k]
```

Comment:
It makes no difference, since modern engines will optimize it away in either case.

<img width="641" alt="Screen Shot 2022-05-17 at 10 45 59" src="https://user-images.githubusercontent.com/1578238/168853484-b4b4c718-c6da-4b52-9953-cd3ba628346c.png">

Reply:
What website/tool is this performance tester? I tried a few whilst making this PR, but they were all a bit terrible.